subsys/bluetooth/host/hci_core.c

```diff
@@ -324,11 +326,37 @@ int bt_hci_cmd_send_sync(uint16_t opcode, struct net_buf *buf,
 
 	LOG_DBG("buf %p opcode 0x%04x len %u", buf, opcode, buf->len);
 
+	/* This local sem is just for suspending the current thread until the
+	 * command is processed by the LL. It is given (and we are awaken) by
+	 * the cmd_complete/status handlers.
+	 */
 	k_sem_init(&sync_sem, 0, 1);
 	cmd(buf)->sync = &sync_sem;
 
 	net_buf_put(&bt_dev.cmd_tx_queue, net_buf_ref(buf));
+	bt_tx_irq_raise();
+
+	/* TODO: disallow sending sync commands from syswq altogether */
+
+	/* Since the commands are now processed in the syswq, we cannot suspend
+	 * and wait. We have to send the command from the current context.
+	 */
+	if (k_current_get() == &k_sys_work_q.thread) {
+		/* drain the command queue until we get to send the command of interest. */
+		struct net_buf *cmd = NULL;
+
+		do {
+			cmd = k_fifo_peek_head(&bt_dev.cmd_tx_queue);
+			LOG_DBG("process cmd %p want %p", cmd, buf);
+
+			bool success = process_pending_cmd(HCI_CMD_TIMEOUT);
 
+			BT_ASSERT_MSG(success, "command opcode 0x%04x timeout", opcode);
+			(void)success; /* cmon zephyr fix your assert macros */
```

Comment:
Seems like `__ASSERT_EVAL()` would be more fitting here (same in the other place I saw you discussing this with @Thalley)

Reply:
in those other places we'd need a `__ASSERT_EVAL_NO_MSG()`.

Reply:
Actually, `__ASSERT_EVAL()` is plain weird. Do I really have to write `__ASSERT_EVAL(success, NULL, "command opcode 0x%04x timeout", opcode);`? Maybe I'm missing something, but to me that's confusing.

Reply:
The way I read it, is that you'd use it as

```	
	__ASSERT_EVAL((void)process_pending_cmd(HCI_CMD_TIMEOUT), 
		      bool success = process_pending_cmd(HCI_CMD_TIMEOUT), success,
		      "command opcode 0x%04x timeout", opcode);
```

Reply:
Example of it being used today is

```c
	__ASSERT_EVAL((void)settings_subsys_init(), int err = settings_subsys_init(), !err,
		      "Failed to initialize settings");
```